include/nbl/asset/IAssetManager.h

```diff
@@ -235,30 +244,61 @@ class IAssetManager : public core::IReferenceCounted, public core::QuitSignallin
             {
-                _override->insertAssetIntoCache(asset, filename, ctx, _hierarchyLevel);
+                _override->insertAssetIntoCache(bundle, filename, ctx, _hierarchyLevel);
             }
-            else if (asset.getContents().empty())
+            else if (bundle.getContents().empty())
             {
                 bool addToCache;
-                asset = _override->handleLoadFail(addToCache, file, filename, filename, ctx, _hierarchyLevel);
-                if (!asset.getContents().empty() && addToCache)
-                    _override->insertAssetIntoCache(asset, filename, ctx, _hierarchyLevel);
+                bundle = _override->handleLoadFail(addToCache, file, filename, filename, ctx, _hierarchyLevel);
+                if (!bundle.getContents().empty() && addToCache)
+                    _override->insertAssetIntoCache(bundle, filename, ctx, _hierarchyLevel);
+            }
+
+            if (bundle.getContents().empty() /*|| none of req levels is dummy*/)
+                return bundle;
+
+            if (restoreLevels)
+            {
+                auto reloadParams = _params;
+                {
+                    using flags_t = std::underlying_type_t<IAssetLoader::E_CACHING_FLAGS>;
+                    constexpr uint32_t bitdepth = sizeof(flags_t)*8u;
+                    const flags_t zeroOutMask = (~static_cast<flags_t>(0)) >> (bitdepth - 2u*(restoreLevels + _hierarchyLevel));
+                    flags_t reloadFlags = reloadParams.cacheFlags;
+                    reloadFlags &= zeroOutMask; // make sure we never pointlessy reload levels above (_restoreLevels+_hierLevel) in reload pass
+                    // set flags for levels [_hierLevel,_hierLevel+_restoreLevels) to dont look into cache and dont put into cache
+                    reloadFlags = core::bitfieldInsert<flags_t>(reloadFlags, IAssetLoader::ECF_DUPLICATE_REFERENCES, _hierarchyLevel*2u, restoreLevels*2u);
+                    reloadParams.cacheFlags = static_cast<IAssetLoader::E_CACHING_FLAGS>(reloadFlags);
+
+                    reloadParams.restoreLevels = 0u; // make sure it wont turn into infinite recursion
+                    reloadParams.reload = true; // TODO (consider): alternative to this flag: another method in override just to let user choose asset for restore
+                }
+                IAssetLoader::SAssetLoadContext ctx(params, file);
+                auto asset = _override->chooseDefaultAsset(bundle, ctx);
+                if (!asset->isAnyDependencyDummy(restoreLevels - 1u))
+                    return bundle;
+
+                auto reloadBundle = getAssetInHierarchy(_file, _supposedFilename, reloadParams, _hierarchyLevel, _override);
+
+                bool restoreSucceeded = _override->handleRestore(std::move(asset), bundle, reloadBundle, restoreLevels);
+                if (!restoreSucceeded) // hm? return empty bundle if restore was requested, but did not succeeded? or just return the bundle? (TODO consider)
```

Comment:
the other `handleRestore` (whole bundle variant), also dont return a bool, check that none of the assets are dummy in the range that interests you.

Return empty bundle, OR return a bundle which only has the assets which were sucessfully restored?

Reply:
(hmm, then maybe in that case the override could choose the assets for restore and construct a new bundle via a non-overridable method which actually checks if restore was successful before adding them to the new bundle)

However if we returned a custom bundle, we'd have a problem because we kinda expect that the `SAssetBundle` we get back from the `getAsset` is the same one that is inside the cache multimap...


So its a bad idea...


Return empty bundle then.